python_modules/dagster/dagster/_core/definitions/asset_subset.py

```diff
@@ -150,56 +187,32 @@ def inverse(
                     dynamic_partitions_store=dynamic_partitions_store,
                 )
             )
-            return self._replace(value=value).as_valid
-
-    def _valid_empty_subset(self, valid: "ValidAssetSubset") -> "ValidAssetSubset":
-        return valid._replace(
-            # unfortunately, this is the best way to get an empty subset of an unknown type
-            # if you don't have access to the partitions definition
-            value=(valid.subset_value - valid.subset_value) if valid.is_partitioned else False
-        )
+            return self._replace(value=value)
 
     def _oper(self, other: "ValidAssetSubset", oper: Callable) -> "ValidAssetSubset":
         value = oper(self.value, other.value)
-        return other._replace(value=value)
+        return self._replace(value=value)
 
-    def __sub__(self, other: "ValidAssetSubset") -> "ValidAssetSubset":
+    def __sub__(self, other: AssetSubset) -> "ValidAssetSubset":
         """Returns an AssetSubset representing self - other if they are compatible, otherwise
-        returns an empty subset.
+        returns self.
         """
-        if not self.is_compatible_with(other):
-            return self._valid_empty_subset(valid=other)
-
-        if not self.is_partitioned:
-            return other._replace(value=self.bool_value and not other.bool_value)
-        return self._oper(other, operator.sub)
-
-    def __rsub__(self, other: "ValidAssetSubset") -> "ValidAssetSubset":
-        """Returns an AssetSubset representing other - self if they are compatible, otherwise
-        returns other.
-        """
-        if not self.is_compatible_with(other):
-            return other
-
+        valid_other = self.get_valid(other)
         if not self.is_partitioned:
-            return other._replace(value=self.bool_value and not other.bool_value)
-        return self._oper(other, operator.sub)
+            return self._replace(value=self.bool_value and not valid_other.bool_value)
+        return self._oper(valid_other, operator.sub)
 
-    def __and__(self, other: "ValidAssetSubset") -> "ValidAssetSubset":
+    def __and__(self, other: AssetSubset) -> "ValidAssetSubset":
         """Returns the intersection of this AssetSubset and another AssetSubset if they are compatible,
```

Comment:
Is there a way to rewrite these comment on these operators in a way that makes it clear how they can be interpreted as acting on sets?

Reply:
@sryza updated!